File: LeetCode/src/39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        results = []
        def dfs(elements: List[int], start: int, k: int) -> None:
            if k <= 0:
                results.append(elements[:]) if k == 0 else None
                return
            
            for idx in range(start, len(candidates)):
                elements.append(candidates[idx])
                dfs(elements, idx, k - candidates[idx])
                elements.pop()
        dfs([], 0, target)
        return results

File: LeetCode/src/39.js
var combinationSum = function(candidates, target) {
    const result = [];
    
    function dfs(elements, start, k) {
        if (k <= 0) {
            if (k == 0) result.push(elements.slice());
            return;
        }
        
        for (let i = start; i < candidates.length; i++) {
            elements.push(candidates[i]);
            dfs(elements, i, k - candidates[i]);
            elements.pop();
        }
    }
    
    dfs([], 0, target);
    return result;
};
